### PR TITLE
Hook to slot stream instead of block stream on the VC

### PR DESCRIPTION
**Pull request by potuz**
Do not requests blocks from the beacon node but rather just slots to track when the head has changed.

The reason why we want to switch is that it will enable us to make the REST VC better. Currently in the REST implementation in `Recv()` we poll for the head block every second and use the resulting block to notify about changes to the head. The problem with this is that if we're optimistically synced, the node will not return a full head block, it can only return a blinded one. This results in a flood of errors in the output. Polling for a blinded block instead could work here (we only care about the slot), but because `validatorClient.StreamBlocks` has to comply with the interface shared between gRPC and HTTP, we would have to create a dummy full block and fill it with at least the slot from the obtained blinded block (the slot is the only thing that matters to the VC). Instead of such hacks, having a slot stream allows us to use Beacon API events in `Recv()`: subscribe to head updates, get the slot from the event and return it inside `StreamSlotsResponse`.

So in short, instead of fetching a blinded block but returning a dummy full block to fix the current optimistic sync issue, we switch over to using slots.

### CODE REVIEW

File: beacon-chain/rpc/prysm/v1alpha1/validator/blocks.go [hunk 1]
```diff
@@ -46,6 +47,53 @@ func (vs *Server) StreamBlocksAltair(req *ethpb.StreamBlocksRequest, stream ethp
...
+			} else {
+				if ev.Type != blockfeed.ReceivedBlock {
+					continue
+				}
+				data, ok := ev.Data.(*blockfeed.ReceivedBlockData)
+				if !ok || data == nil {
+					continue
+				}
+				s = data.SignedBlock.Block().Slot()
+			}
```
Comment (terencechain) [inline beacon-chain/rpc/prysm/v1alpha1/validator/blocks.go:83]: Is there a use case for non verified slot received? is this just for beacon api?

File: validator/client/validator.go [hunk 2]
```diff
@@ -317,58 +315,35 @@ func (v *validator) WaitForSync(ctx context.Context) error {
...
-			blk, err = blocks.NewSignedBeaconBlock(b.Phase0Block)
-		case *ethpb.StreamBlocksResponse_AltairBlock:
-			blk, err = blocks.NewSignedBeaconBlock(b.AltairBlock)
-		case *ethpb.StreamBlocksResponse_BellatrixBlock:
-			blk, err = blocks.NewSignedBeaconBlock(b.BellatrixBlock)
-		case *ethpb.StreamBlocksResponse_CapellaBlock:
-			blk, err = blocks.NewSignedBeaconBlock(b.CapellaBlock)
-		case *ethpb.StreamBlocksResponse_DenebBlock:
-			blk, err = blocks.NewSignedBeaconBlock(b.DenebBlock)
+		if res.Slot > v.highestSlot() {
```
Comment (terencechain) [inline validator/client/validator.go:343]: You could encapsulate `if res.Slot > v.highestSlot() ` inside `v.setHighestSlot`, since the naming already suggested that you will only set the slot if it's the highest

File: beacon-chain/rpc/prysm/v1alpha1/validator/blocks.go (hunk 1 above)
Comment (potuz) [reply beacon-chain/rpc/prysm/v1alpha1/validator/blocks.go:83]: no, this is just legacy code, I didn't feel like removing the field to keep it as close as it was before, but we can remove it if needed.

File: validator/client/validator.go (hunk 2 above)
Comment (potuz) [reply validator/client/validator.go]: thanks, we can also encapsulate the send

**Review by terencechain — DISMISSED** [review]

**Review by terencechain — APPROVED** [review]

File: proto/prysm/v1alpha1/validator.proto [hunk 3]
```diff
@@ -294,6 +294,16 @@ service BeaconNodeValidator {
         };
     }
 
+    // Server-side stream of all slots of valid blocks as they are received by
+    // the beacon chain node.
+    // DEPRECATED: This endpoint is superseded by the /eth/v1/events Beacon API endpoint
+    rpc StreamSlots(StreamSlotsRequest) returns (stream StreamSlotsResponse) {
```
Comment (james-prysm) [inline proto/prysm/v1alpha1/validator.proto:300]: In the future if we want to enable people to use our validator with other clients... is there something else we can use in its place? it's still marked as deprecated

Comment (potuz) [reply proto/prysm/v1alpha1/validator.proto:300]: yeah, the REST validator should hook to `eth/v1/events`  and use the BeaconAPI. This PR is a placeholder in the REST case.

**Review by james-prysm — APPROVED** [review]

File: validator/client/service.go [hunk 4]
```diff
@@ -226,16 +225,6 @@ func (v *ValidatorService) Start() {
 		validatorRegBatchSize:          v.validatorRegBatchSize,
 	}
 
-	// To resolve a race condition at startup due to the interface
```
Comment (prestonvanloon) [inline validator/client/service.go:229]: Nice to see this hack can be removed.

**Review by prestonvanloon — APPROVED** [review]
Thanks for the description. This all makes sense and LGTM

**Comment by prestonvanloon** [comment]
Labeling as breaking change since this isn't compatible with the prior release of beacon node. The beacon node will need to be updated before the validator in the next release.
